### PR TITLE
cli: use the channel-map api for status

### PR DESCRIPTION
**Pull request by sergiusens**
Also:

- introduce --experimental-progressive-delivery
- fixup the channel-map schema

Signed-off-by: Sergio Schvezov <sergio.schvezov@canonical.com>

- [ ] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [ ] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [ ] Have you successfully run `./runtests.sh static`?
- [ ] Have you successfully run `./runtests.sh tests/unit`?

-----

### CODE REVIEW

File: snapcraft/cli/_channel_map.py [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+
+from snapcraft.storeapi.v2.channel_map import (
+    ChannelMap,
+    MappedChannel,
+    Revision,
+    SnapChannel,
+)
+
+
+_HINTS_T = NamedTuple(
```
Comment (cjp256) [inline snapcraft/cli/_channel_map.py:31]: I'd suggest using mypy to for constant checks with `Final`.

Either something like `_SYMBOL_CLOSED: Final[str] = "-"`
Or If you prefer, bundle them into a class:
```
from typing import Final
class Hints:
  CLOSED: Final[str] = "-"
  FOLLOWING: Final[str] = "↑"
...
```

File: snapcraft/cli/store.py [hunk 2]
```diff
@@ -360,19 +362,39 @@ def close(snap_name, channels):
...
     """Get the status on the store for <snap-name>.
 
     \b
     Examples:
         snapcraft status my-snap
-        snapcraft status my-snap --arch armhf
     """
-    snapcraft.status(snap_name, arch)
+    if experimental_progressive_delivery:
+        os.environ["SNAPCRAFT_EXPERIMENTAL_PROGRESSIVE_DELIVERY"] = "Y"
```
Comment (cjp256) [inline snapcraft/cli/store.py:383]: This is already handled by apply_host_provider_flags(), but I have used strtobool, so they could be made consistent to 0|1 for boolean flags for checking throughout code.

Comment (sergiusens) [reply snapcraft/cli/store.py:383]: We discussed this and agreed that this only applies to build commands (this is not one) and that one solution would be to implement a click.Option that does this for us.

File: snapcraft/cli/_channel_map.py (hunk 1 above)
Comment (sergiusens) [reply snapcraft/cli/_channel_map.py:31]: Sure.

File: snapcraft/cli/_channel_map.py [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+    if revision is not None:
+        version_string = revision.version
+        revision_string = f"{revision.revision}"
+
+    if mapped_channel is not None:
+        if channel_info.branch is None and mapped_channel.progressive.percentage:
+            channel_string = ""
+        elif channel_info.branch is not None:
+            channel_string = f"{channel_info.risk}/{channel_info.branch}"
+            # This value can be None.
```
Comment (cjp256) [inline snapcraft/cli/_channel_map.py:102]: Can probably drop the comment if check is:
`if mapped_channel.expiration_date is not None`

File: snapcraft/cli/_channel_map.py [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+) -> bool:
+    for channel_name in channels:
+        for progressive in (False, True):
+            try:
+                snap_channel_map.get_mapped_channel(
+                    channel_name=channel_name,
+                    architecture=architecture,
+                    progressive=progressive,
+                )
+                return True
```
Comment (cjp256) [inline snapcraft/cli/_channel_map.py:212]: uh oh, here's a case of multiple returns ;)

File: snapcraft/cli/_channel_map.py [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+            except ValueError:
```
Comment (cjp256) [inline snapcraft/cli/_channel_map.py:213]: I wonder if get_mapped_channel() should have returned a None.

File: snapcraft/storeapi/v2/_api_schema.py [hunk 6]
```diff
@@ -198,7 +198,9 @@
                                 "type": ["string", "null"],
                             },
                         },
-                        "required": ["name", "creation-date", "pattern"],
+                        # pattern is documented as required but is not returned,
```
Comment (cjp256) [inline snapcraft/storeapi/v2/_api_schema.py:201]: So the API is not passing their own schema?  Is the store team aware?

Comment (sergiusens) [reply snapcraft/storeapi/v2/_api_schema.py:201]: yes, they are fixing it. I suspect (and hope) this is a bug in the documentation and not the schema itself

File: snapcraft/cli/_channel_map.py (hunk 5 above)
Comment (sergiusens) [reply snapcraft/cli/_channel_map.py:213]: we have been value based more than exception based so far in our duo team endeavor which might strike a cord with familiarity.

Let's discuss this later when we have more time and adapt our coding style to reflect that.

File: snapcraft/cli/_channel_map.py (hunk 3 above)
Comment (sergiusens) [reply snapcraft/cli/_channel_map.py:102]: done

File: snapcraft/cli/_channel_map.py (hunk 4 above)
Comment (sergiusens) [reply snapcraft/cli/_channel_map.py:212]: fixed

File: snapcraft/cli/_channel_map.py [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+    if revision is not None:
+        version_string = revision.version
+        revision_string = f"{revision.revision}"
+
+    if mapped_channel is not None:
+        if channel_info.branch is None and mapped_channel.progressive.percentage:
+            channel_string = ""
+        elif channel_info.branch is not None:
+            channel_string = f"{channel_info.risk}/{channel_info.branch}"
+            # This value can be None.
```
Comment (cjp256) [inline snapcraft/cli/_channel_map.py:99]: rm this comment?

**Comment by sergiusens** [comment]
the spread failure is related to spread mixing up the new test in master for go.mod into the tests for this snap. Ignoring failure for this reason

**Pull request merged**